### ISSUE TITLE
linter: support no-unnecessary-conditions rule

### ISSUE THREAD

**Issue by copilotenstar** [oxc-project/oxc#18287]
One of the few rules that catches actual bugs.

Recently in elysiaJS:

```ts				
let decoded = true <--------- 
for (let i = 0; i < secrets.length; i++) {
	if (typeof value !== 'string') throw new InvalidCookieSignature(name)
	const temp = await unsignCookie(value, secrets[i])

	if (temp !== false) {
		decoded = true
		value = temp

		break
	}
}

if (!decoded) throw new InvalidCookieSignature(name)
```

linter could've caught this bug and prevented a security issue. 

to be fair, i'm surprised it's not caught by typescript already. [they do catch some cases though](https://i.imgur.com/1qniQHm.png)

**Comment by connorshea** [comment]
This is a type-aware rule and so will be implemented via tsgolint, this rule is being tracked in https://github.com/oxc-project/tsgolint/issues/188
